File: code_samples/section12/example_5_exponential_search/exponential_search.py
"""Lesson 6 - Exponential Search (with step counting and file-based tests)."""

from __future__ import annotations
import os
import math

# pylint: disable=broad-exception-caught
# pylint: disable=line-too-long


def exponential_search(arr: list[int], target: int) -> tuple[int, int]:
    """
    Perform exponential search on a sorted array.

    Exponential search works in two main phases:
      1) Exponential range expansion:
         Repeatedly double an index (1, 2, 4, 8, ...) until the target
         is less than or equal to arr[bound], or the bound exceeds the array.
      2) Binary search:
         Once a candidate range is identified, perform binary search
         within that range.

    Returns:
        (index, step_count)
        - index: position of target in arr, or -1 if not found
        - step_count: approximate number of comparisons performed

    Assumes:
        - `arr` is sorted in ascending order
    """
    steps = 0                # Counts the number of comparison steps
    n = len(arr)             # Number of elements in the array

    # Handle empty array case
    if n == 0:
        return -1, steps

    # --------------------------------------------------
    # Step 0: Check the first element directly
    # --------------------------------------------------
    steps += 1
    if arr[0] == target:
        return 0, steps

    # --------------------------------------------------
    # Phase 1: Find a range [bound/2, bound] where target may lie
    # --------------------------------------------------
    bound = 1

    # Expand the bound exponentially while it is within array limits
    while bound < n:
        steps += 1  # compare arr[bound] == target
        if arr[bound] == target:
            return bound, steps

        steps += 1  # compare arr[bound] < target or > target
        if arr[bound] > target:
            break   # upper bound found; stop expanding

        bound *= 2  # double the bound and continue searching

    # --------------------------------------------------
    # Phase 2: Binary search within the identified range
    # --------------------------------------------------
    left = bound // 2
    right = min(bound, n - 1)

    # Standard binary search within [left, right]
    while left <= right:
        mid = left + (right - left) // 2

        steps += 1  # compare arr[mid] == target
        if arr[mid] == target:
            return mid, steps

        steps += 1  # compare arr[mid] < target
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    # Target not found
    return -1, steps


def load_ordered_file() -> list[int] | None:
    """
    Attempt to load ordered.txt from several likely relative locations.

    This allows the script to be run from different directories, such as:
      - the lesson folder
      - section12/
      - the repository root

    Returns:
        A list of integers if the file is successfully loaded,
        or None if all attempts fail.
    """
    print(f"Working directory: {os.getcwd()}")

    # Candidate relative paths to ordered.txt
    candidates = [
        os.path.join("..", "data", "ordered.txt"),                # from lesson_6_exponential_search/
        os.path.join("..", "..", "data", "ordered.txt"),          # from section12/
        os.path.join("code_samples", "section12", "data", "ordered.txt"),
    ]

    # Try each candidate path
    for rel in candidates:
        full = os.path.abspath(rel)
        print(f"Attempting to read: {full}")

        if os.path.exists(full):
            try:
                with open(full, "r", encoding="utf-8") as f:
                    text = f.read()

                # Split on any whitespace and parse integers
                nums = [int(tok) for tok in text.split()]
                print(f"Loaded {len(nums)} integers from {full}")
                return nums

            except Exception as ex:  # noqa: BLE001
                # File exists but could not be read or parsed
                print(f"Error reading ordered.txt:\n{ex}")
        else:
            # File does not exist at this location
            print("  File does not exist here.")

    # All candidate paths failed
    print("Missing input file - aborting.")
    return None


def run_tests(arr: list[int]) -> None:
    """
    Run a set of exponential-search tests and print results.

    The output includes:
      - array size
      - approximate log2(n)
      - number of steps required for each search case
    """
    n = len(arr)
    approx_log_n = math.log2(n) if n > 0 else 0.0
    print(f"\nArray size n = {n}, log2(n) ≈ {approx_log_n:.2f}\n")

    # Test values:
    #   - first element
    #   - middle element
    #   - last element
    #   - value not present
    tests = [
        arr[0],               # first element
        arr[n // 2],          # middle element
        arr[-1],              # last element
        6001,                 # not present (assuming data is -5000..5000)
    ]

    # Run exponential search for each test value
    for t in tests:
        idx, steps = exponential_search(arr, t)
        if idx != -1:
            print(f"Target {t:6} found at index {idx:5}, steps = {steps:4}")
        else:
            print(f"Target {t:6} not found,            steps = {steps:4}")


if __name__ == "__main__":
    # Load ordered data and run tests if successful
    data = load_ordered_file()
    if data is not None:
        run_tests(data)

File: code_samples/section12/example_5_exponential_search/exponential_search.java
package code_samples.section12.example_5_exponential_search;

import java.io.File;                 // Represents file and directory pathnames
import java.io.FileNotFoundException; // Thrown when a file cannot be found
import java.nio.file.Files;          // File existence checks
import java.nio.file.Path;           // File path abstraction
import java.nio.file.Paths;          // Path construction utilities
import java.util.ArrayList;           // Dynamic list for file loading
import java.util.List;
import java.util.Scanner;             // Token-based file reading

public class exponential_search {

    // ----------------------------------------------------
    // Result holder for search index + step count
    // ----------------------------------------------------

    /**
     * Simple immutable container used to return both:
     *  - index: where the target was found (or -1)
     *  - steps: number of array probes performed
     *
     * This avoids using out-parameters and keeps the API clear.
     */
    static class Result {
        final int index;
        final int steps;

        Result(int index, int steps) {
            this.index = index;
            this.steps = steps;
        }
    }

    // ----------------------------------------------------
    // Exponential search with step counting
    // ----------------------------------------------------

    /**
     * Performs exponential search on a sorted integer array.
     *
     * Exponential search operates in two phases:
     *
     *  1) Exponential range expansion:
     *     Starting from index 1, repeatedly double the index
     *     (1, 2, 4, 8, ...) until either:
     *       - the target is less than or equal to arr[bound], or
     *       - the bound exceeds the array size.
     *
     *  2) Binary search:
     *     Once a candidate range is found, perform binary search
     *     within that range.
     *
     * Step counting:
     *   - Counts how many times the algorithm probes the array
     *     (i.e., reads arr[i]).
     *
     * @param arr     Sorted array of integers
     * @param target  Value to locate
     *
     * @return Result object containing index and step count
     */
    static Result exponentialSearchWithSteps(int[] arr, int target) {
        int n = arr.length;
        int steps = 0;

        // Handle empty array
        if (n == 0) {
            return new Result(-1, steps);
        }

        // ------------------------------------------------
        // Step 0: Check the first element directly
        // ------------------------------------------------
        int val0 = arr[0];
        steps++;                   // One array probe
        if (val0 == target) {
            return new Result(0, steps);
        }

        // ------------------------------------------------
        // Phase 1: Find range by repeated doubling
        // ------------------------------------------------
        int bound = 1;

        // Expand the bound until:
        //  - bound exceeds array length, or
        //  - arr[bound] >= target
        while (bound < n) {
            int v = arr[bound];
            steps++;               // One array probe

            if (v < target) {
                bound *= 2;        // Exponentially increase range
            } else {
                break;             // Found upper bound
            }
        }

        // Determine the final search interval
        int left = bound / 2;
        int right = Math.min(bound, n - 1);

        // ------------------------------------------------
        // Phase 2: Binary search within [left, right]
        // ------------------------------------------------
        return binarySearchRangeWithSteps(arr, left, right, target, steps);
    }

    // ----------------------------------------------------
    // Binary search helper with step counting
    // ----------------------------------------------------

    /**
     * Performs binary search on a subrange [lo, hi] of the array,
     * continuing the step count accumulated during exponential search.
     *
     * @param arr        Sorted array of integers
     * @param lo         Lower bound index (inclusive)
     * @param hi         Upper bound index (inclusive)
     * @param target     Value to locate
     * @param stepsSoFar Step count accumulated so far
     *
     * @return Result object with final index and step count
     */
    static Result binarySearchRangeWithSteps(
            int[] arr,
            int lo,
            int hi,
            int target,
            int stepsSoFar) {

        int steps = stepsSoFar;

        // Standard binary search loop
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;

            int val = arr[mid];
            steps++;               // One array probe

            if (val == target) {
                return new Result(mid, steps);
            } else if (val < target) {
                lo = mid + 1;      // Search right half
            } else {
                hi = mid - 1;      // Search left half
            }
        }

        // Target not found in this range
        return new Result(-1, steps);
    }

    // ----------------------------------------------------
    // File loading helpers
    // ----------------------------------------------------

    /**
     * Attempts to load ordered.txt from multiple candidate paths.
     *
     * This makes the program robust when run from different
     * working directories (lesson folder vs workspace root).
     *
     * Candidate paths:
     *   1) ../data/ordered.txt
     *   2) code_samples/section12/data/ordered.txt
     *
     * @return Array of integers if found; otherwise null
     */
    static int[] loadOrderedFile() {
        String[] candidates = new String[] {
            ".." + File.separator + "data" + File.separator + "ordered.txt",
            "code_samples" + File.separator + "section12"
                + File.separator + "data"
                + File.separator + "ordered.txt"
        };

        // Print current working directory for diagnostics
        System.out.println("Working directory: " + System.getProperty("user.dir"));

        // Try each candidate path
        for (String rel : candidates) {
            Path p = Paths.get(rel).normalize().toAbsolutePath();
            System.out.println("Trying: " + p);

            if (Files.exists(p)) {
                try (Scanner sc = new Scanner(new File(p.toString()))) {
                    List<Integer> list = new ArrayList<>();

                    // Read integers from file
                    while (sc.hasNextInt()) {
                        list.add(sc.nextInt());
                    }

                    // Convert List<Integer> → int[]
                    int[] arr = new int[list.size()];
                    for (int i = 0; i < list.size(); i++) {
                        arr[i] = list.get(i);
                    }

                    System.out.println("Loaded " + arr.length
                            + " integers from " + p);
                    return arr;

                } catch (FileNotFoundException e) {
                    // Should not occur after exists(), but guarded anyway
                    System.out.println(
                        "Unexpected FileNotFound after exists() for: " + p);
                }
            } else {
                System.out.println("  Not found at: " + p);
            }
        }

        // All candidate paths failed
        System.out.println(
            "Missing input file - could not locate ordered.txt via any candidate path.");
        return null;
    }

    // ----------------------------------------------------
    // Demo / tests
    // ----------------------------------------------------

    /**
     * Program entry point.
     *
     * Loads ordered data from disk and runs exponential search
     * on representative target values, printing index and step count.
     */
    public static void main(String[] args) {
        int[] arr = loadOrderedFile();
        if (arr == null) {
            // Diagnostics already printed
            return;
        }

        // Representative test targets:
        int[] testTargets = new int[] {
            -5000,   // first element (best case)
            0,       // middle element
            5000,    // last element
            6000     // value not present
        };

        // Run exponential search for each target
        for (int target : testTargets) {
            Result res = exponentialSearchWithSteps(arr, target);
            System.out.printf(
                "Target %d -> index = %d, steps = %d%n",
                target, res.index, res.steps);
        }
    }
}
